components/selectors/matching.rs

```diff
@@ -466,7 +452,10 @@ fn matches_simple_selector<E>(
         }
         SimpleSelector::Negation(ref negated) => {
             !negated.iter().all(|s| {
-                matches_complex_selector(s, element, parent_bf, relations, flags)
+                match matches_complex_selector_internal(s, element, relations, flags) {
```

Comment:
We might have descendant combinators inside the `:not(...)`.  Don't we want to be able to use the ancestor filter to quickly return false for them too?